components/modal-clubs.jsx

```diff
@@ -0,0 +1,295 @@
+var React = require('react/addons');
+var Select = require('react-select');
+var _ = require('underscore');
+var Modal = require('../components/modal.jsx');
+var ModalManagerMixin = require('../mixins/modal-manager');
+var TeachAPIClientMixin = require('../mixins/teach-api-client');
+var LoginLink = require('../components/login.jsx').LoginLink;
+var Map = require('../components/map.jsx');
+
+var normalizeClub = function(clubState) {
+  var state = _.extend({}, clubState);
+
+  if (state.website && !/^https?:\/\//.test(state.website)) {
+    state.website = 'http://' + state.website;
+  }
+
+  return state;
+};
+
+var validateClub = function(clubState) {
+  var errors = [];
+
+  if (!clubState.name) {
+    errors.push("You must provide a name for your club.");
+  }
+  if (!clubState.description) {
+    errors.push("You must provide a description for your club.");
+  }
+  if (!clubState.location) {
+    errors.push("You must provide a location for your club.");
+  }
+  if (clubState.website && !/\./.test(clubState.website)) {
+    errors.push("Your club's website must be a valid URL.");
+  }
+
+  return errors;
+};
+
+var ModalAddOrChangeYourClub = React.createClass({
+  mixins: [React.addons.LinkedStateMixin, ModalManagerMixin,
+           TeachAPIClientMixin],
+  contextTypes: {
+    router: React.PropTypes.func
+  },
+  propTypes: {
+    // If club is provided, then we're a 'change' dialog, otherwise
+    // we're an 'add' dialog.
+    club: React.PropTypes.object,
+    idPrefix: React.PropTypes.string,
+    onSuccess: React.PropTypes.func.isRequired
+  },
+  getDefaultProps: function() {
+    return {
+      idPrefix: 'ModalAddOrChangeYourClub_'
+    };
+  },
+  statics: {
+    normalizeClub: normalizeClub,
+    validateClub: validateClub,
+    teachAPIEvents: {
+      'username:change': 'handleUsernameChange',
+    }
+  },
+  STEP_AUTH: 1,
+  STEP_FORM: 2,
+  STEP_WAIT_FOR_NETWORK: 3,
+  STEP_SHOW_RESULT: 4,
+  getInitialState: function() {
+    var clubState = {
+      name: '',
+      website: '',
+      description: '',
+      location: '',
+      latitude: null,
+      longitude: null
+    };
+    if (this.props.club) {
+      _.extend(clubState, _.pick(this.props.club,
```

Comment:
if we allow ES6, then this already works as `Object.assign` without a need for underscore